File: android/app/src/main/kotlin/com/operon/app/shell/InfoSheet.kt
package com.operon.app.shell

import android.content.Context
import android.graphics.Color
import android.graphics.Typeface
import android.graphics.drawable.GradientDrawable
import android.util.TypedValue
import android.view.Gravity
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.google.android.material.progressindicator.LinearProgressIndicator
import com.operon.app.R
import com.operon.app.shell.SheetUi.color
import com.operon.app.shell.SheetUi.dp

/**
 * Read-only stats (ios/App/App/InfoSheet.swift) — the context-window
 * breakdown and the Claude subscription quotas both render through this.
 *
 * One card per section: an optional header line with a value and a progress
 * bar, then plain rows, then an optional footer. Values are monospaced because
 * they are numbers the user compares down a column.
 */
object InfoSheet {

    data class Row(
        val label: String,
        val value: String?,
        val detail: String?,
        /** `#rrggbb`, drawn as a swatch before the label. */
        val color: String?,
        val indent: Boolean,
    )

    enum class Tone {
        NORMAL, WARN, ERROR;

        companion object {
            fun from(raw: String?): Tone = when (raw) {
                "warn" -> WARN
                "error" -> ERROR
                else -> NORMAL
            }
        }
    }

    data class Section(
        val header: String?,
        val value: String?,
        /** 0…1 */
        val progress: Double?,
        val tone: Tone,
        val footer: String?,
        val rows: List<Row>,
    )

    fun present(
        context: Context,
        title: String,
        caption: String?,
        sections: List<Section>,
        onDismiss: () -> Unit,
    ): BottomSheetDialog {
        val body = SheetUi.page(context) {
            addView(
                SheetUi.title(context, title).apply {
                    if (caption != null) setPadding(context.dp(4), context.dp(4), context.dp(4), 0)
                },
                SheetUi.matchWidth(),
            )
            if (caption != null) {
                addView(
                    TextView(context).apply {
                        text = caption.uppercase()
                        setTextSize(TypedValue.COMPLEX_UNIT_SP, 11f)
                        setTypeface(typeface, Typeface.BOLD)
                        setTextColor(context.color(R.attr.operonMutedForeground))
                        setPadding(context.dp(4), 0, context.dp(4), context.dp(12))
                    },
                    SheetUi.matchWidth(),
                )
            }

            sections.forEachIndexed { index, section ->
                if (index > 0) addView(SheetUi.spacer(context))
                addView(sectionCard(context, section), SheetUi.matchWidth())
            }
        }

        return SheetUi.present(context, body, onDismiss)
    }

    private fun sectionCard(context: Context, section: Section): LinearLayout = SheetUi.card(context) {
        setPadding(context.dp(14), context.dp(12), context.dp(14), context.dp(12))

        if (section.header != null || section.value != null) {
            addView(
                LinearLayout(context).apply {
                    orientation = LinearLayout.HORIZONTAL
                    gravity = Gravity.CENTER_VERTICAL
                    addView(
                        SheetUi.label(context, section.header.orEmpty(), sizeSp = 14f).apply {
                            layoutParams = LinearLayout.LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT)
                                .apply { weight = 1f }
                        },
                    )
                    section.value?.let {
                        addView(mono(context, it, toneColor(context, section.tone)))
                    }
                },
                SheetUi.matchWidth(),
            )
        }

        section.progress?.let { progress ->
            addView(
                LinearProgressIndicator(context).apply {
                    max = 1000
                    setProgressCompat((progress.coerceIn(0.0, 1.0) * 1000).toInt(), false)
                    trackCornerRadius = context.dp(3)
                    trackThickness = context.dp(6)
                    setIndicatorColor(barColor(context, section.tone))
                    trackColor = context.color(R.attr.operonTile)
                    layoutParams = SheetUi.matchWidth().apply { topMargin = context.dp(8) }
                },
            )
        }

        section.rows.forEachIndexed { index, row ->
            addView(
                rowView(context, row).apply {
                    layoutParams = SheetUi.matchWidth().apply {
                        topMargin = if (index == 0) context.dp(8) else context.dp(6)
                    }
                },
            )
        }

        section.footer?.let { footer ->
            addView(
                TextView(context).apply {
                    text = footer
                    setTextSize(TypedValue.COMPLEX_UNIT_SP, 11f)
                    setTextColor(context.color(R.attr.operonMutedForeground))
                    gravity = Gravity.END
                    layoutParams = SheetUi.matchWidth().apply { topMargin = context.dp(8) }
                },
            )
        }
    }

    private fun rowView(context: Context, row: Row): LinearLayout = LinearLayout(context).apply {
        orientation = LinearLayout.HORIZONTAL
        gravity = Gravity.CENTER_VERTICAL

        parseHex(row.color)?.let { swatch ->
            addView(
                android.view.View(context).apply {
                    background = GradientDrawable().apply {
                        cornerRadius = context.dp(2).toFloat()
                        setColor(swatch)
                    }
                    layoutParams = LinearLayout.LayoutParams(context.dp(8), context.dp(8))
                        .apply { marginEnd = context.dp(8) }
                },
            )
        }

        addView(
            SheetUi.label(context, row.label, sizeSp = 13f, colorId = R.attr.operonMutedForeground).apply {
                layoutParams = LinearLayout.LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT).apply {
                    weight = 1f
                    marginStart = if (row.indent) context.dp(12) else 0
                }
            },
        )

        row.value?.let {
            addView(
                mono(
                    context,
                    it,
                    context.color(if (row.indent) R.attr.operonMutedForeground else R.attr.operonForeground),
                ),
            )
        }
        row.detail?.let {
            addView(
                mono(context, it, context.color(R.attr.operonMutedForeground)).apply {
                    minWidth = context.dp(44)
                    gravity = Gravity.END
                    layoutParams = LinearLayout.LayoutParams(
                        ViewGroup.LayoutParams.WRAP_CONTENT,
                        ViewGroup.LayoutParams.WRAP_CONTENT,
                    ).apply { marginStart = context.dp(8) }
                },
            )
        }
    }

    private fun mono(context: Context, text: String, colour: Int): TextView = TextView(context).apply {
        this.text = text
        setTextSize(TypedValue.COMPLEX_UNIT_SP, 13f)
        typeface = Typeface.MONOSPACE
        setTextColor(colour)
        maxLines = 1
    }

    private fun toneColor(context: Context, tone: Tone): Int = when (tone) {
        Tone.NORMAL -> context.color(R.attr.operonForeground)
        Tone.WARN -> context.color(R.attr.operonStatusWarn)
        Tone.ERROR -> context.color(R.attr.operonStatusError)
    }

    private fun barColor(context: Context, tone: Tone): Int = when (tone) {
        Tone.NORMAL -> context.color(R.attr.operonBrand)
        Tone.WARN -> context.color(R.attr.operonStatusWarn)
        Tone.ERROR -> context.color(R.attr.operonStatusError)
    }

    /** `#rrggbb` / `#rgb`; null for anything else (CSS variables, hsl()…). */
    private fun parseHex(value: String?): Int? {
        val raw = value?.trim()?.removePrefix("#") ?: return null
        val hex = when (raw.length) {
            3 -> raw.map { "$it$it" }.joinToString("")
            6 -> raw
            else -> return null
        }
        return try {
            Color.parseColor("#$hex")
        } catch (_: IllegalArgumentException) {
            null
        }
    }
}

File: ios/App/App/InfoSheet.swift
import SwiftUI
import UIKit

/// Read-only stats as a system sheet — the context-window breakdown and the
/// Claude subscription quotas both render through this. One card per
/// section: an optional header line (label + value) with a progress bar, then
/// plain rows (label / value / percent, an optional color swatch), then an
/// optional footer line.
struct InfoSheetSection: Identifiable {
    struct Row: Identifiable {
        let id = UUID()
        let label: String
        let value: String?
        let detail: String?
        /// `#rrggbb`, drawn as a swatch before the label.
        let color: String?
        let indent: Bool
    }
    enum Tone: String { case normal, warn, error }
    /// Bar fill level, separate from `tone` so a bar can go green → orange → red
    /// while its label stays quiet until the limit is close.
    enum BarTone: String { case ok, warn, error }

    let id = UUID()
    let header: String?
    let value: String?
    /// 0…1
    let progress: Double?
    let tone: Tone
    /// nil: the bar follows `tone`.
    let barTone: BarTone?
    let footer: String?
    let rows: [Row]
}

struct InfoSheetView: View {
    let title: String
    let caption: String?
    let sections: [InfoSheetSection]

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 12) {
                VStack(alignment: .leading, spacing: 2) {
                    Text(title)
                        .font(.title3.weight(.semibold))
                    if let caption {
                        Text(caption)
                            .font(.caption.weight(.medium))
                            .foregroundStyle(.secondary)
                            .textCase(.uppercase)
                    }
                }
                .padding(.horizontal, 4)
                .padding(.top, 4)

                ForEach(sections) { section in
                    VStack(alignment: .leading, spacing: 8) {
                        if section.header != nil || section.value != nil {
                            HStack(alignment: .firstTextBaseline) {
                                Text(section.header ?? "")
                                    .font(.subheadline.weight(.medium))
                                Spacer(minLength: 8)
                                if let value = section.value {
                                    Text(value)
                                        .font(.system(.footnote, design: .monospaced))
                                        .foregroundStyle(tint(for: section))
                                }
                            }
                        }
                        if let progress = section.progress {
                            ProgressView(value: min(max(progress, 0), 1))
                                .tint(barTint(for: section))
                        }
                        if !section.rows.isEmpty {
                            VStack(spacing: 6) {
                                ForEach(section.rows) { row in
                                    HStack(spacing: 8) {
                                        if let hex = row.color, let color = Color(hex: hex) {
                                            RoundedRectangle(cornerRadius: 2).fill(color).frame(width: 8, height: 8)
                                        }
                                        Text(row.label)
                                            .font(.footnote)
                                            .foregroundStyle(row.indent ? Color.secondary.opacity(0.8) : Color.secondary)
                                            .padding(.leading, row.indent ? 12 : 0)
                                            .lineLimit(1)
                                        Spacer(minLength: 8)
                                        if let value = row.value {
                                            Text(value)
                                                .font(.system(.footnote, design: .monospaced))
                                                .foregroundStyle(row.indent ? Color.secondary : Color.primary)
                                        }
                                        if let detail = row.detail {
                                            Text(detail)
                                                .font(.system(.footnote, design: .monospaced))
                                                .foregroundStyle(.secondary)
                                                .frame(minWidth: 44, alignment: .trailing)
                                        }
                                    }
                                }
                            }
                            .padding(.top, section.header == nil && section.progress == nil ? 0 : 2)
                        }
                        if let footer = section.footer {
                            HStack {
                                Spacer()
                                Text(footer)
                                    .font(.caption)
                                    .foregroundStyle(.secondary)
                            }
                        }
                    }
                    .padding(.horizontal, 14)
                    .padding(.vertical, 12)
                    .background(Color(.secondarySystemGroupedBackground), in: RoundedRectangle(cornerRadius: 14, style: .continuous))
                }
            }
            .padding(.horizontal, 16)
            .padding(.top, 20)
            .padding(.bottom, 16)
        }
        .background(Color(.systemGroupedBackground))
    }

    private func tint(for section: InfoSheetSection) -> Color {
        switch section.tone {
        case .normal: return .primary
        case .warn: return .orange
        case .error: return .red
        }
    }

    private func barTint(for section: InfoSheetSection) -> Color {
        if let barTone = section.barTone {
            switch barTone {
            case .ok: return fillOk
            case .warn: return fillWarn
            case .error: return fillError
            }
        }
        switch section.tone {
        case .normal: return brand
        case .warn: return .orange
        case .error: return .red
        }
    }
}

/// Fill tokens from globals.css (light / dark): `--color-accent-green`,
/// `--color-accent-warm`, `--color-destructive` — the colors the web bars use.
private func dynamicFill(light: UInt32, dark: UInt32) -> Color {
    func color(_ hex: UInt32) -> UIColor {
        UIColor(red: CGFloat((hex >> 16) & 0xFF) / 255,
                green: CGFloat((hex >> 8) & 0xFF) / 255,
                blue: CGFloat(hex & 0xFF) / 255,
                alpha: 1)
    }
    return Color(UIColor { trait in trait.userInterfaceStyle == .dark ? color(dark) : color(light) })
}

private let fillOk = dynamicFill(light: 0x3DB87A, dark: 0x4FCC8E)
private let fillWarn = dynamicFill(light: 0xE5845C, dark: 0xF09570)
private let fillError = dynamicFill(light: 0xEF4444, dark: 0x991B1B)

extension Color {
    /// `#rrggbb` / `#rgb`; nil for anything else (CSS variables, hsl()…).
    init?(hex: String) {
        var s = hex.trimmingCharacters(in: .whitespaces)
        guard s.hasPrefix("#") else { return nil }
        s.removeFirst()
        if s.count == 3 { s = s.map { "\($0)\($0)" }.joined() }
        guard s.count == 6, let v = UInt32(s, radix: 16) else { return nil }
        self.init(
            red: Double((v >> 16) & 0xFF) / 255,
            green: Double((v >> 8) & 0xFF) / 255,
            blue: Double(v & 0xFF) / 255
        )
    }
}

extension InfoSheetView {
    static func present(from presenter: UIViewController, title: String, caption: String?, sections: [InfoSheetSection], onDismiss: @escaping () -> Void) {
        let view = InfoSheetView(title: title, caption: caption, sections: sections)
        let height: CGFloat = NativeSheet.titleHeight + (caption == nil ? 0 : 18) + sections.reduce(0) { acc, s in
            acc + 24 + (s.header != nil || s.value != nil ? 22 : 0) + (s.progress != nil ? 12 : 0)
                + CGFloat(s.rows.count) * 24 + (s.footer != nil ? 20 : 0) + 12
        }
        NativeSheet.present(from: presenter, content: view, contentHeight: height, dismisser: SheetDismisser(), onDismiss: onDismiss)
    }
}
